pkg/frontend/querymiddleware/cardinality_query_cache.go

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/cache"
+	"github.com/grafana/dskit/tenant"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/querier/cardinality"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+const (
+	cardinalityLabelNamesQueryCachePrefix  = "cn:"
+	cardinalityLabelValuesQueryCachePrefix = "cv:"
+)
+
+type cardinalityQueryRequest interface {
+	// String returns a full representation of the request, used to uniquely identify
+	// a request in the cache.
+	String() string
+
+	// RequestType returns the type of the cardinality query, used to distinguish
+	// different type of requests in the cache.
+	RequestType() cardinality.RequestType
+}
+
+// cardinalityQueryCache is a http.RoundTripped wrapping the downstream with an HTTP response cache.
+// This RoundTripper is used to add caching support to cardinality analysis API endpoints.
+type cardinalityQueryCache struct {
+	cache  cache.Cache
+	limits Limits
+	next   http.RoundTripper
+	logger log.Logger
+}
+
+func newCardinalityQueryCacheRoundTripper(cache cache.Cache, limits Limits, next http.RoundTripper, logger log.Logger) http.RoundTripper {
+	return &cardinalityQueryCache{
+		cache:  cache,
+		limits: limits,
+		next:   next,
+		logger: logger,
+	}
+}
+
+// TODO metrics?
+func (c *cardinalityQueryCache) RoundTrip(req *http.Request) (*http.Response, error) {
+	ctx := req.Context()
+
+	spanLog, ctx := spanlogger.NewWithLogger(ctx, c.logger, "cardinalityQueryCache.RoundTrip")
+	defer spanLog.Finish()
+
+	// Skip the cache if disabled for this request.
+	if decodeCacheDisabledOption(req) {
+		level.Debug(spanLog).Log("msg", "cache disabled for the request")
+		return c.next.RoundTrip(req)
+	}
+
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	// Skip the cache if disabled for the tenant.
+	cacheTTL := validation.SmallestPositiveNonZeroDurationPerTenant(tenantIDs, c.limits.ResultsCacheTTLForCardinalityQuery)
```

Comment:
nit: If tenantIDs has two tenants and one of them has caching disabled, but other one doesn't, this will return cacheTTL from tenant that doesn't have it disabled. That seems like a bug to me.

Reply:
Good point. It's not a nit, but a bug! Will add a test to cover it too.

Reply:
Test added in https://github.com/grafana/mimir/pull/5212/commits/768a4a5a508add98e1d6d69589ef82f80cc0bd00.